src/main/java/uk/gov/hmcts/ccd/data/casedetails/DefaultCaseDetailsRepository.java

```diff
@@ -75,9 +77,16 @@ public CaseDetails set(final CaseDetails caseDetails) {
             throw new CaseConcurrencyException("The case data has been altered outside of this transaction.");
         } catch (PersistenceException e) {
             LOG.warn("Failed to store case details", e);
-            if (e.getCause() instanceof ConstraintViolationException) {
-                throw new CaseConcurrencyException(e.getMessage());
+
+            if (e.getCause() instanceof ConstraintViolationException
+                && ((ConstraintViolationException) e.getCause()).getConstraintName()
+                .equals(UNIQUE_REFERENCE_KEY_CONSTRAINT)) {
+
+                LOG.warn("ConstraintViolationException happen for UUID={}. ConstraintName: {}",
+                    caseDetails.getReference(), UNIQUE_REFERENCE_KEY_CONSTRAINT);
+                throw new ReferenceKeyUniqueConstraintException(e.getMessage());
             }
+            throw new CasePersistenceException(e.getMessage());
```

Comment:
note a change of behaviour here. Before we were just logging non `ConstraintViolationException` PersistenceExceptions, now we are throwing them. But it seems better throwing them to me than swallowing them so 👍

Reply:
We can throw new Exceptions with the original one kept in the "cause".